src/ngraph_assign_clusters.cc

```diff
@@ -83,23 +91,203 @@ namespace {
+      src_predicate != dst_predicate) {
+    is_deadness_ok = false;
+    return Status::OK();
+  }
+
+  // Case src X , dst True // invalid scenario
+  // If src has Non-True Predicate and dst has True Predicate, it implies that
+  // the dst node is control flow
+  if (!DeadnessAnalysis::IsTruePredString(src_predicate) &&
+      DeadnessAnalysis::IsTruePredString(dst_predicate)) {
+    return errors::Internal("Attempting to cluster control-flow node ",
+                            dst->name(), "[", dst->type_string(), "]");
+  }
+
+  // Case src True, dst Y
+  // If this edge is contracted, all the outputs of the merged cluster will
+  // have the predicate Y. Hence contraction is possible only when, all
+  // outputs of the src cluster (other than the current edge) have the
+  // predicate Y
+  if (DeadnessAnalysis::IsTruePredString(src_predicate)) {
+    auto src_cluster_out_edges = cluster_map[src]->outgoing_edges;
+    bool found_same_out_preds = true;
+    std::string pred_check = dst_predicate;
+
+    for (const Edge* src_cluster_edge : src_cluster_out_edges) {
+      if (src_cluster_edge == edge) {
+        continue;
+      }
+      Node* src_cluster_dst = src_cluster_edge->dst();
+      std::string src_cluster_dst_pred =
+          cluster_map[src_cluster_dst]->predicate_string;
+      if (pred_check != src_cluster_dst_pred) {
+        found_same_out_preds = false;
+        break;
+      }
+    }
+    // Cannot contract this edge
+    if (!found_same_out_preds) {
+      is_deadness_ok = false;
+      return Status::OK();
+    }
+  }
+
+  // Case src X, dst Y, X==Y
+  // Case src True, dst Y(Y can be True), all other output edges from src
+  // cluster have pred Y
+  // Ok to contract
+  is_deadness_ok = true;
+  return Status::OK();
 }
 
+// Some sanity checks for Node's cluster assignment wrt Deadness
+Status CheckNodeClusterAssignmentWRTDeadness(
+    Node* node, std::map<Node*, string>& nodes_predicate_map,
+    string& cluster_pred_string) {
+  if (nodes_predicate_map.find(node) == nodes_predicate_map.end()) {
+    return errors::Internal("Node ", node->name(), " [", node->type_string(),
+                            "]", " not found in predicate map");
+  }
+  std::string node_pred_string = nodes_predicate_map[node];
+
+  if (DeadnessAnalysis::IsControlFlowPredString(node_pred_string)) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]",
+        " should not be clustered as it is a control flow op");
+  }
+
+  if (!DeadnessAnalysis::IsTruePredString(node_pred_string) &&
+      node_pred_string != cluster_pred_string) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]", " Predicate : ",
+        node_pred_string,
+        "should not be clustered in cluster with pred_String ",
+        cluster_pred_string);
+  }
+  return Status::OK();
+}
+#endif
+
+// Merges src and dst clusters of the edge
+void MergeClusters(Edge* edge,
```

Comment:
maybe add a warning. Only call this when you are sure that X/Y/Z checks have passed (in the sense that this function does not perform any checks)